### PR TITLE
[Brick Game] fix bouncing of the ball on the player

### PR DESCRIPTION
**Pull request by nicosemp**
fix: the ball now bounces off the player at an angle
- the angle is based on where the player hits the ball
- the vector velocity is always 1
- the ball still starts towards bottom right at 45deg

fix: the ball now inverts x/y direction instead of setting to 1 or -1
- this is to keep the bouncing anglefix checkCollisionWithBricks logic
- it used to pass through some bricks instead of changing direction

fix: missing parenthesis in the Brick's fill color switch statement

p.s. sorry about all the edits I'm still learning github PRs

### CODE REVIEW

File: src/components/views/BrickGameView/Game.ts [hunk 1]
```diff
@@ -284,7 +281,17 @@ class Ball {
       return;
     if (this.position.x + this.size.width < this.app.player.position.x) return;
 
-    this.direction.y = -1; //Moving up now
+    // Maps a value from the "in" range to the "out" range. Maybe this should go in a Mathy utils file?
+    const mapValue = (value: number, in_min: number, in_max: number, out_min: number, out_max: number) => {
+      return (value - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+    }
```
Comment (tvillarete) [inline src/components/views/BrickGameView/Game.ts]:
```suggestion
    const mapValue = (value: number, inMin: number, inMax: number, outMin: number, outMax: number) => {
      return (value - inMin) * (outMax - outMin) / (inMax - inMin) + outMin;
    }
```

File: src/components/views/BrickGameView/Game.ts [hunk 2]
```diff
@@ -352,7 +359,7 @@ class Brick {
         this.app.context.fillStyle = 'rgb(0, 240, 0)'; //Green
         break;
       case 2:
-        this.app.context.fillStyle = 'rgb(255,140,0'; //Orange?
+        this.app.context.fillStyle = 'rgb(255, 140, 0)'; //Orange? Looks orange to me :)
```
Comment (tvillarete) [inline src/components/views/BrickGameView/Game.ts]: Nice catch! Not sure why the question mark was there 😅

```suggestion
        this.app.context.fillStyle = 'rgb(255, 140, 0)'; //Orange
```

**Review by tvillarete — COMMENTED** [review]
This is fantastic – thanks for doing this @nicosemp! One small thing I noticed was that the velocity of the ball is just _slightly_ slower in these new changes. Ideally we should keep the velocity the same as the original. What are your thoughts?

Before            |  After
:-------------------------:|:-------------------------:
before | after

**Comment by nicosemp** [comment]
> This is fantastic – thanks for doing this @nicosemp!

You are very welcome! I really like this project.

> One small thing I noticed was that the velocity of the ball is just slightly slower in these new changes. Ideally we should keep the velocity the same as the original. What are your thoughts?

I thought you might notice this. As you probably know, the length of the vector is the velocity.
In your original file, the velocity is `sqrt(2)` because you define both the X and Y speed to be 1 or -1.

The velocity I used is exactly 1, just to make the math simpler.
I made a quick (and ugly) drawing to demonstrate:

image

In the original version the vector length ends up being `sqrt(2)` (about 1.41), which is fairly larger than 1.

### The solution
These variables are just implicitly multiplied by 1:
```
const vx = Math.cos(angle);
const vy = Math.sin(angle);
```
So if you want to keep the original speed you can just multiply by `sqrt(2)` instead of 1:
```
const vx = Math.cos(angle) * Math.sqrt(2);
const vy = Math.sin(angle) * Math.sqrt(2);
```
In case you do, remember to also change the starting speed both in the Ball class `constructor` and in its `reset` function!

**Comment by tvillarete** [comment]
Thanks for the great explanation! Seems simple enough to change later on, so I think we can leave the speed you've set in this PR for now and I'll file a followup PR with some velocity tweaks (and maybe a victory screen? 😄)

Again, thanks a bunch for these improvements!

**Comment by nicosemp** [comment]
It was a pleasure! Thank you for merging :)

**Pull request merged**